### ISSUE TITLE
Linux 5.2.x breaks VM start - qubes-db fails to start

### ISSUE THREAD

**Issue by marmarek** [QubesOS/qubes-issues#5199]
**Qubes OS version**
4.0

**Affected component(s) or functionality**
Linux / vchan

**Brief summary**
With update to Linux 5.2.x, qubes-db service fails to start with `gnttab: error: mmap failed: No such device or address` error.

**To Reproduce**
_Steps to reproduce the behavior:_
1. Install kernel-latest-qubes-vm-5.2.3
2. Switch a VM to that kernel, make sure `virt_mode` is set to `pvh`
3. Try to start the VM

**Expected behavior**
VM starts normally

**Actual behavior**
VM either crash on start (report on IRC by @xaki23), or just qubes-db fails to start leaving VM half-configured.
Logs from service start, with xen_gntdev debugging enabled:
```
Jul 28 17:18:49 localhost systemd[1]: Started Cleanup of Temporary Directories.
Jul 28 18:02:35 localhost systemd[1]: Starting Qubes DB agent...
Jul 28 18:02:35 localhost kernel: xen:xen_gntdev: priv 00000000689318be
Jul 28 18:02:35 localhost kernel: xen:xen_gntdev: priv 00000000689318be, add 1
Jul 28 18:02:35 localhost kernel: xen:xen_gntdev: map 0+1 at 725a8e4ad000 (pgoff 0)
Jul 28 18:02:35 localhost kernel: xen:xen_gntdev: map 0+1
Jul 28 18:02:35 localhost kernel: xen:xen_gntdev: priv 00000000689318be, add 1
Jul 28 18:02:35 localhost kernel: xen:xen_gntdev: map 1+1 at 725a8e47b000 (pgoff 1)
Jul 28 18:02:35 localhost kernel: xen:xen_gntdev: map 1+1
Jul 28 18:02:35 localhost kernel: xen:xen_gntdev: priv 00000000689318be, del 4096+1
Jul 28 18:02:35 localhost kernel: xen:xen_gntdev: unmap 1+1 [0+1]
Jul 28 18:02:35 localhost kernel: xen:xen_gntdev: unmap handle=257 st=0
Jul 28 18:02:35 localhost kernel: xen:xen_gntdev: priv 00000000689318be, offset for vaddr 725a8e4ad000
Jul 28 18:02:35 localhost kernel: xen:xen_gntdev: gntdev_vma_close 000000007cc48ca0
Jul 28 18:02:35 localhost kernel: xen:xen_gntdev: priv 00000000689318be, del 0+1
Jul 28 18:02:35 localhost kernel: xen:xen_gntdev: unmap 0+1 [0+1]
Jul 28 18:02:35 localhost kernel: xen:xen_gntdev: unmap handle=256 st=0
Jul 28 18:02:35 localhost kernel: xen:xen_gntdev: priv 00000000689318be
Jul 28 18:02:35 localhost qubesdb-daemon[843]: gnttab: error: mmap failed: No such device or address
Jul 28 18:02:35 localhost qubesdb-daemon[843]: FATAL: vchan initialization failed
Jul 28 18:02:35 localhost systemd[1]: qubes-db.service: Main process exited, code=exited, status=1/FAILURE
Jul 28 18:02:35 localhost systemd[1]: qubes-db.service: Failed with result 'exit-code'.
Jul 28 18:02:35 localhost systemd[1]: Failed to start Qubes DB agent.
```
strace fragment:
```
[pid   857] openat(AT_FDCWD, "/dev/xen/evtchn", O_RDWR|O_CLOEXEC) = 8
[pid   857] ioctl(8, IOCTL_EVTCHN_BIND_INTERDOMAIN, 0x7ffd3407b778) = 32
[pid   857] write(8, " \0\0\0", 4)      = 4
[pid   857] ioctl(7, IOCTL_GNTDEV_MAP_GRANT_REF, 0x7ffd3407b6d0) = 0
[pid   857] mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_SHARED, 7, 0) = 0x777f1211b000
[pid   857] ioctl(7, IOCTL_GNTDEV_SET_UNMAP_NOTIFY, 0x7ffd3407b710) = 0
[pid   857] ioctl(7, IOCTL_GNTDEV_MAP_GRANT_REF, 0x7ffd3407b6d0) = 0
[pid   857] mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_SHARED, 7, 0x1000) = -1 ENXIO (No such device or address)
[pid   857] write(2, "gnttab: ", 8)     = 8
[pid   857] write(2, "error: ", 7)      = 7
[pid   857] write(2, "mmap failed", 11) = 11
[pid   857] write(2, ": No such device or address", 27) = 27
[pid   857] write(2, "\n", 1)           = 1
[pid   857] ioctl(7, IOCTL_GNTDEV_UNMAP_GRANT_REF, 0x7ffd3407b710) = 0
[pid   857] ioctl(7, IOCTL_GNTDEV_GET_OFFSET_FOR_VADDR, 0x7ffd3407b750) = 0
[pid   857] munmap(0x777f1211b000, 4096) = 0
[pid   857] ioctl(7, IOCTL_GNTDEV_UNMAP_GRANT_REF, 0x7ffd3407b740) = 0
[pid   857] close(8)                    = 0
[pid   857] close(7)                    = 0
```

cc @m-v-b

**Comment by marmarek** [comment]
As @m-v-b found, it's most likely caused by https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?id=df9bde015a72ffd978e39a750662c7cf579b1715

I suspect it's about `vma->vm_pgoff` usage - previously it was used only for finding map index, now `vm_map_pages` uses it for starting offset in `map->pages` too.
If I understand this code correctly, gntdev driver abuses offset mmap argument for an grant index (multiplied by page size) instead of real offset. I think the proper solution is to revert this commit.

**Comment by m-v-b** [comment]
Hello again,

My kernel build has finally successfully finished. I carried a few basic boot-up tests with AppVMs, and I confirm that reverting the commit at https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?id=df9bde015a72ffd978e39a750662c7cf579b1715 resolves this issue.

Please note that I did not use the kernel I built with `dom0` yet. I will test that in a moment.

**Comment by marmarek** [comment]
Relevant discussion: https://lore.kernel.org/lkml/CAFqt6zZN+6r6wYJY+f15JAjj8dY+o30w_+EWH9Vy2kUXCKSBog@mail.gmail.com/

@m-v-b could you test replacing `vm_map_pages` with `vm_map_pages_zero` instead of the revert?

**Comment by m-v-b** [comment]
@marmarek, of course, I do not mind helping with that. Please note that I need about 12 to 16 hours before I can report back with the test results, mostly due to my employment/work. If such a time frame would be too late, please let me know.

**Comment by marmarek** [comment]
I'm testing it already.

**Comment by marmarek** [comment]
It works.

**Comment by marmarek** [comment]
Fix queued in 5.2-stable already.

**Comment by DemiMarie** [comment]
@marmarek would it be possible to push out a new `kernel-latest-qubes-vm` package that includes the fix?

**Comment by marmarek** [comment]
As soon as it will be released (5.2.7), which should happen this week.

**Comment by marmarek** [comment]
kernel-latest-qubes-vm-5.2.7 is in current-testing: https://github.com/QubesOS/updates-status/issues/1248
